### ISSUE TITLE
v2.1 Missing media queries with SSR

### ISSUE THREAD

**Issue by 0x80** [styled-components/styled-components#939]
In 2.1.0 (compared to 2.0.0), some of my media queries have broken in production. Something simple like below doesn't trigger and I get no error. It looks like it's just missing from the CSS.

```css
 @media (min-width: ${breakpoints.sm}) {
    width: ${imageSizeLg};
    height: ${imageSizeLg};
  }
```

I'm using Gatsby, so production meaning SSR. This is the [code integrating styled-components](https://github.com/gatsbyjs/gatsby/blob/1.0/packages/gatsby-plugin-styled-components/src/gatsby-ssr.js)

What I find strange is that there __are__ still media query statements in the generated CSS, even ones that are using the exact same query.

Any idea what might have caused this?

**Comment by k15a** [comment]
Is this just happening with server-side rendering or also when you use styled-components on the client? If this is also happening on the client could you share a small reproduction at codesandbox or webpackbin?

**Comment by 0x80** [comment]
This happens only with SSR. Things work during development with client side rendering.

**Comment by philpl** [comment]
cc @geelen

**Comment by geelen** [comment]
Hmm, that's super super weird. We bumped a stylis version between those right @philpl?

@0x80 could you post the generated code somewhere I could take a look at?

**Comment by philpl** [comment]
@geelen Yea, we did actually. @0x80 Have you verified that this only happens in combination with SSR? Can you just paste component code into a webpackbin, or codesandbox, so we can see if it's related to SSR at all?

**Comment by 0x80** [comment]
I've switched about 3 times now between 2.0 and 2.1. It's only the built SSR output that has the problem. I'm about to leave for a festival for a few days. I'll try to isolate something when I get back next week.

**Comment by Cinamonas** [comment]
If I can be of any help, this works fine:

```js
const Block = styled.div`
  display: none;

  @media (min-width: 62.5em) {
    display: block;
  }
`;
```
```scss
// Generated CSS:
.eFENwo {display: none;}@media (min-width: 62.5em) {.eFENwo {display: block;}}
```

While this doesn't:

```js
const mediaQuery = (...args) => css`
  @media (max-width: 62.5em) {
    ${css(...args)}
  }
`;

const Block = styled.div`
  display: none;

  ${mediaQuery`
    display: block;
  `}
`;
```
```scss
// Generated CSS:
.iyQpre {display: none;}.iyQpre@media (max-width: 62.5em) {.iyQpre {
    display: block;
  ;}}
```

Notice that `@media` is prefixed with styled component's class name, which is why this case fails.

**Comment by 0x80** [comment]
I am trying to mimic SSR in a webpackbin, but I'm not sure how to get it working. If I call `new ServerStyleSheet()` I get an error `Uncaught Error: BrowserTag cannot be cloned!`

Here's the code: https://www.webpackbin.com/bins/-KnFSs12stbx0kw07H5V

**Comment by mschipperheyn** [comment]
There is a repo I set up for another issue I have w styled components (https://github.com/styled-components/styled-components/issues/762). It works with SSR. Maybe it's useful to you. https://github.com/mschipperheyn/react-universally 

There are a couple of branches on this repo related to styled-components

**Comment by 0x80** [comment]
@mschipperheyn Thanks for the tip. In the meantime I was setting up this quick experiment:

https://github.com/0x80/styled-ssr

I tried to match the approach that Gatsby uses for SSR but I can't reproduce the issue yet 😞

**Comment by philpl** [comment]
I was able to reproduce this:

```css
.djurMT {background-color: hotpink;width: 100px;height: 100px;border-radius: 3px;-webkit-flex: none;-ms-flex: none;flex: none;overflow: hidden;-webkit-transform: translate3d(0, 0, 0);-ms-transform: translate3d(0, 0, 0);transform: translate3d(0, 0, 0);}.djurMT@media (max-width: 62.5em) {.djurMT {
    width: ;
    height: ;
  ;}}
```

The relevant part being:

```css
.djurMT@media
```

**But** I was only able to reproduce it using `preprocess` turned on in the babel plugin. So at least @Cinamonas reproduction is only a configuration mistake. Do note that `preprocess` is experimental and just a test leading up to v3, if you've purposefully turned it on :wink:

I'm also sure that this triggers @0x80 bug, due to an edge-case in the preprocessing:

```js
[["@media (min-width: ", breakpoints.sm, ") {"], [" {width: ", imageSizeLg, ";height: ", imageSizeLg, ";}}"]]
```

The preprocessing maps over the first level of arrays (the second being for the interpolations) and prepends each of them with the classname. When we use a media query, the first set of rules is not supposed to have a classname prepended to it (duh), so it fails. If you do want to continue using the experimental preprocessing, just put a dummy rule in front of the media query.

And last but not least:

```js
const mediaQuery = (...args) => css`
  @media (max-width: 62.5em) {
    ${css(...args)}
  }
`;

mediaQuery`display: block`
```

This will **never** work with preprocessing, as we can't preprocess dynamic input. The correct usage with preprocessing would be:

```js
const mediaQuery = x => css`
  @media (max-width: 62.5em) {
    ${x}
  }
`;

mediaQuery(css`display: block`)
```

I hope this explains the issue :wink:

**Comment by 0x80** [comment]
@philpl Awesome! Ah yes, the babel plugin...I forgot I was using that. I'm not sure what you mean by preprocessing. My config for production is `{"ssr": true,"displayName": true}`.

**Comment by philpl** [comment]
@0x80 that's certainly really peculiar. Can you show me some output of your transpilation that is exhibiting this bug please? Seems that there's still sth weird afoot

**Comment by 0x80** [comment]
@philpl I tried to replicate it by adding the same babel plugins to my [test](https://github.com/0x80/styled-ssr) but I can't reproduce it there. There must be some important difference in the way Gatsby handles SSR compared to my simplified approach.

What I __can__ tell you is that the issue disappears when I remove the styled-components babel plugin.  Here's the difference in css output

```javascript
const CoverImage = styled.div`
  width: ${imageSizeSm};
  height: ${imageSizeSm};
  border-radius: 3px;
  flex: none;
  overflow: hidden;
  transform: translate3d(0, 0, 0);
  color: hotpink;

  @media (min-width: ${breakpoints.sm}) {
    width: ${imageSizeLg};
    height: ${imageSizeLg};
  }
`
```

Without plugin:
```css
/* sc-component-id: sc-kjoXOD */

.dfQzLK{width: 100px;height: 100px;border-radius: 3px;-webkit-flex: none;-ms-flex: none;flex: none;overflow: hidden;-webkit-transform: translate3d(0, 0, 0);-ms-transform: translate3d(0, 0, 0);transform: translate3d(0, 0, 0);color: hotpink;}@media (min-width: 40em){.dfQzLK{width: 140px;height: 140px;}}
```

With plugin:
```css
/* sc-component-id: work-preview__CoverImage-s1qj064d-2 */

.dfqqrO{width: 100px;height: 100px;border-radius: 3px;-webkit-flex: none;-ms-flex: none;flex: none;overflow: hidden;}
```

**Comment by geelen** [comment]
Wow that is super strange! Removing the `transform` line doesn't help does it?

**Comment by 0x80** [comment]
@geelen Actually, above that line there is a block comment. I didn't include it in the example because I assumed it was irrelevant.

Turns out if I remove the comment the code works 😮 

This was the full source:

```javascript
const CoverImage = styled.div`
  width: ${imageSizeSm};
  height: ${imageSizeSm};
  border-radius: 3px;
  flex: none;
  overflow: hidden;

  /*
    We have to pull this element to the GPU otherwise clipping the circle
    doesn't work during the animation of blur on Safari
  */
  transform: translate3d(0, 0, 0);
  color: hotpink;

  @media (min-width: ${breakpoints.sm}) {
    width: ${imageSizeLg};
    height: ${imageSizeLg};
  }
`
```

**Comment by probablyup** [comment]
Yeah known bug with the babel plugin since it uses an old version of
stylis.

**Comment by geelen** [comment]
Ooh sounds like a result! I'll leave this to @philpl hopefully we can cut a new release of the parser and we'll be golden.

**Comment by philpl** [comment]
If it's fixed with removing that comment I'll just put out a patch version in a minute.

Some issues in this thread however are caused by preprocessing. So if there are still issues after the release, please disable the preprocess option and if there's still a bug open a new issue with your babel output and try turning the minify option off for a temporary fix.

**Comment by philpl** [comment]
I've published a new version of the babel plugin, so the original bug should be resolved.

If anyone here finds another issue only when the babel plugin is enabled, please open an issue on the babel-plugin-styled-components repo.
